### PR TITLE
[Impeller] started asserting golden test runner has fatal impeller validations

### PR DESCRIPTION
**Pull request by gaaclarke**
fixes https://github.com/flutter/flutter/issues/145017

This works by removing the conditional compilation for validation and turning them to be fatal in the test runner's main.

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide] and the [C++, Objective-C, Java style guides].
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I added new tests to check the change I am making or feature I am adding, or the PR is [test-exempt]. See [testing the engine] for instructions on writing and running engine tests.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I signed the [CLA].
- [x] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[test-exempt]: https://github.com/flutter/flutter/wiki/Tree-hygiene#tests
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[C++, Objective-C, Java style guides]: https://github.com/flutter/engine/blob/main/CONTRIBUTING.md#style
[testing the engine]: https://github.com/flutter/flutter/wiki/Testing-the-engine
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat

### CODE REVIEW

File: impeller/golden_tests/main.cc [hunk 1]
```diff
@@ -24,7 +25,14 @@ void print_usage() {
 }
 }  // namespace
 
+namespace impeller {
+TEST(ValidationTest, IsFatal) {
+  EXPECT_TRUE(ImpellerValidationErrorsAreFatal());
```
Comment (gaaclarke) [inline impeller/golden_tests/main.cc:30]: I tried to use `EXPECT_EXIT` but I couldn't get it to stop killing the process.

Comment (jonahwilliams) [reply impeller/golden_tests/main.cc:30]: You want to use EXPECT_DEATH? 

That seems to work for me

**Review by matanlurey — APPROVED** [review]
LGTM, but agree with `expect_death` if able.

**Comment by gaaclarke** [comment]
> LGTM, but agree with `expect_death` if able.

I'll give that another try today.

**Comment by gaaclarke** [comment]
Yea looks like EXPECT_DEATH is unreliable in processes with more than one thread:
```
[ RUN      ] ValidationTest.IsFatal

[WARNING] ../../flutter/third_party/googletest/googletest/src/gtest-death-test.cc:1102:: Death tests use fork(), which is unsafe particularly in a threaded context. For this test, Google Test detected 6 threads. See https://github.com/google/googletest/blob/main/docs/advanced.md#death-tests-and-threads for more explanation and suggested solutions, especially if this is the last message you see before your test times out.
[FATAL:flutter/impeller/base/validation.cc(57)] Break on 'ImpellerValidationBreak' to inspect point of failure: 
Abort trap: 6
```

I looked to see if I could make sure the test happens first, presumably before the other threads are spawned, and it doesn't look like there is a way to make that the case reliably with googletest.

**Pull request merged**